reference/test/preconditioner/isai_kernels.cpp

```diff
@@ -0,0 +1,626 @@
+{
+    using Csr = typename TestFixture::Csr;
+    using value_type = typename TestFixture::value_type;
+    auto result = this->clone_allocations(gko::lend(this->l_csr));
+
+    gko::kernels::reference::isai::generate_l_inverse(
+        this->exec, gko::lend(this->l_csr), gko::lend(result));
+
+    GKO_ASSERT_MTX_EQ_SPARSITY(result, this->l_csr_inv);
+    GKO_ASSERT_MTX_NEAR(result, this->l_csr_inv, r<value_type>::value);
+}
+
+
+TYPED_TEST(Isai, KernelGenerateL2)
+{
+    using Csr = typename TestFixture::Csr;
+    using value_type = typename TestFixture::value_type;
+    const auto l_mtx = this->transpose(gko::lend(this->u_csr));
+    auto result = this->clone_allocations(gko::lend(l_mtx));
+
+    gko::kernels::reference::isai::generate_l_inverse(
+        this->exec, gko::lend(l_mtx), gko::lend(result));
+
+    const auto expected = this->transpose(gko::lend(this->u_csr_inv));
+    GKO_ASSERT_MTX_EQ_SPARSITY(result, expected);
+    GKO_ASSERT_MTX_NEAR(result, expected, r<value_type>::value);
+}
+
+
+TYPED_TEST(Isai, KernelGenerateLsparse1)
+{
+    using Csr = typename TestFixture::Csr;
+    using value_type = typename TestFixture::value_type;
+    auto result = this->clone_allocations(gko::lend(this->l_sparse));
+
+    gko::kernels::reference::isai::generate_l_inverse(
+        this->exec, gko::lend(this->l_sparse), gko::lend(result));
+
+    GKO_ASSERT_MTX_EQ_SPARSITY(result, this->l_sparse_inv);
+    GKO_ASSERT_MTX_NEAR(result, this->l_sparse_inv, r<value_type>::value);
+}
+
+
+TYPED_TEST(Isai, KernelGenerateLsparse2)
+{
+    using Csr = typename TestFixture::Csr;
+    using value_type = typename TestFixture::value_type;
+    using index_type = typename TestFixture::index_type;
+    auto result = this->clone_allocations(gko::lend(this->l_sparse2));
+
+    gko::kernels::reference::isai::generate_l_inverse(
+        this->exec, gko::lend(this->l_sparse2), gko::lend(result));
+
+    GKO_ASSERT_MTX_EQ_SPARSITY(result, this->l_sparse2_inv);
+    GKO_ASSERT_MTX_NEAR(result, this->l_sparse2_inv, r<value_type>::value);
+}
+
+
+TYPED_TEST(Isai, KernelGenerateLsparse3)
+{
+    using Csr = typename TestFixture::Csr;
+    using value_type = typename TestFixture::value_type;
+    using index_type = typename TestFixture::index_type;
+    const auto l_mtx = this->transpose(gko::lend(this->u_sparse));
+    auto result = this->clone_allocations(gko::lend(l_mtx));
+
+    gko::kernels::reference::isai::generate_l_inverse(
+        this->exec, gko::lend(l_mtx), gko::lend(result));
+
+    // Results in a slightly different version than u_sparse_inv->transpose()
+    // because a different row-sparsity pattern is used in u_sparse vs. l_mtx
+    // (only one value changes compared to u_sparse_inv->transpose())
+    const auto expected = Csr::create(
+        this->exec, gko::dim<2>(4, 4),
+        I<value_type>{-.5, .125, .25, .3125, .125, -.5, .125, .25, .5},
+        I<index_type>{0, 0, 1, 0, 1, 2, 0, 2, 3}, I<index_type>{0, 1, 3, 6, 9});
```

Comment:
Same here, are the `I<...>` needed? `Csr` should already have the different types embeded in its template parameters, so I would be surprised.

Reply:
I use it here as an alias to `std::initializer_list<>`. The problem is that braced initialization lists do not work with unspecified template arguments because the type is unknown (therefore, it also does not work with perfect forwarding).
`auto` has a special behavior with braced initialization lists, it will automatically convert it to an `std::initializer_list`. For templates, this is not the case.